src/Compilers/CSharp/Portable/Lowering/Instrumentation/LocalStateTracingInstrumenter.cs

```diff
@@ -0,0 +1,621 @@
+                        MakeStoreLoggerArguments(parameterLogger.Parameters[0], parameter, parameter.Type, _factory.Parameter(parameter), refAssignmentSourceIndex: null, _factory.Literal((ushort)parameter.Ordinal)))));
+                }
+            }
+
+            if (previousPrologue != null)
+            {
+                prologueBuilder.Add(previousPrologue);
+            }
+
+            prologue = _factory.StatementList(prologueBuilder.ToImmutableAndFree());
+
+            var (entryOverload, entryArgs) = (isLambdaBody, isStateMachine) switch
+            {
+                (isLambdaBody: false, isStateMachine: false) =>
+                    (WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogMethodEntry,
+                    new[] { _factory.MethodDefIndex(_factory.TopLevelMethod) }),
+                (isLambdaBody: true, isStateMachine: false) =>
+                    (WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogLambdaEntry,
+                    new[] { _factory.MethodDefIndex(_factory.TopLevelMethod), _factory.MethodDefIndex(_factory.CurrentFunction) }),
+                (isLambdaBody: false, isStateMachine: true) =>
+                    (WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogStateMachineMethodEntry,
+                    new[] { _factory.MethodDefIndex(_factory.TopLevelMethod), _factory.StateMachineInstanceId() }),
+                (isLambdaBody: true, isStateMachine: true) =>
+                    (WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogStateMachineLambdaEntry,
+                    new[] { _factory.MethodDefIndex(_factory.TopLevelMethod), _factory.MethodDefIndex(_factory.CurrentFunction), _factory.StateMachineInstanceId() }),
+            };
+
+            var entryLogger = GetWellKnownMethodSymbol(entryOverload, _factory.Syntax);
+            var instrumentationPrologue = (entryLogger != null) ?
+                _factory.Assignment(_factory.Local(_scope.ContextVariable), _factory.Call(receiver: null, entryLogger, entryArgs)) : _factory.NoOp(NoOpStatementFlavor.Default);
+
+            var returnLogger = GetWellKnownMethodSymbol(WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogReturn, _factory.Syntax);
+            var instrumentationEpilogue = (returnLogger != null) ?
+                _factory.ExpressionStatement(_factory.Call(receiver: _factory.Local(_scope.ContextVariable), returnLogger)) : _factory.NoOp(NoOpStatementFlavor.Default);
+
+            // currently don't need to compose multiple instrumentations
+            Debug.Assert(instrumentation is null);
+
+            instrumentation = new BoundBlockInstrumentation(
+                _factory.Syntax,
+                _scope.ContextVariable,
+                instrumentationPrologue,
+                instrumentationEpilogue);
+
+            _scope.Close(isMethodBody);
```

Comment:
>_scope.Close(isMethodBody) [](http://example.com/codeflow?start=12&length=26)

Am I understanding correctly that there is no explicit `_scope.Open()` for the method body, just for lambdas?

Reply:
Just lambdas and local functions. We start with an open scope for the method body.